Distinguish PATCH and PUT event in handling of BIOS and Installer

**Current situation**
- `PATCH` ~ `HTTP PATCH` 
- `UPDATE` ~ `HTTP PUT`
But the business logic is not separated clearly when handle these event

**Expectation**
In `Bios` and `Installer` event handler
- `PATCH`: update partly database entity properties if `request payload` present these properties
- `UPDATE`: update all database entity properties. Normally, it is required all `request payload` ~ `entity model`

These entity properties are update-able:
- `version`
- `state`
- `deploy_config`
- `published_by`

**Affect sub projects**
  - `:edge:bios`
  - `:edge:core`
  - `:edge:module:installer`

**More information**
- Some error logs when updating

```
org.jooq.exception.DataAccessException: SQL [update "PUBLIC"."TBL_MODULE" set "PUBLIC"."TBL_MODULE"."SERVICE_NAME" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."MODIFIED_AT" = cast(? as timestamp), "PUBLIC"."TBL_MODULE"."CREATED_AT" = cast(? as timestamp), "PUBLIC"."TBL_MODULE"."DEPLOY_ID" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."DEPLOY_CONFIG_JSON" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."VERSION" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."SERVICE_ID" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."STATE" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."DEPLOY_LOCATION" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."SERVICE_TYPE" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."PUBLISHED_BY" = cast(? as varchar) where (1 = 1 and "PUBLIC"."TBL_MODULE"."SERVICE_ID" = cast(? as varchar))]; NULL not allowed for column "CREATED_AT"; SQL statement:
update "PUBLIC"."TBL_MODULE" set "PUBLIC"."TBL_MODULE"."SERVICE_NAME" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."MODIFIED_AT" = cast(? as timestamp), "PUBLIC"."TBL_MODULE"."CREATED_AT" = cast(? as timestamp), "PUBLIC"."TBL_MODULE"."DEPLOY_ID" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."DEPLOY_CONFIG_JSON" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."VERSION" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."SERVICE_ID" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."STATE" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."DEPLOY_LOCATION" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."SERVICE_TYPE" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."PUBLISHED_BY" = cast(? as varchar) where (1 = 1 and "PUBLIC"."TBL_MODULE"."SERVICE_ID" = cast(? as varchar)) [23502-197]
        at org.jooq_3.11.8.H2.debug(Unknown Source)
        at org.jooq.impl.Tools.translate(Tools.java:2384)
        at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:822)
        at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:364)
        at org.jooq.impl.AbstractDelegatingQuery.execute(AbstractDelegatingQuery.java:127)
        at io.github.jklingsporn.vertx.jooq.rx.jdbc.JDBCRXGenericQueryExecutor.lambda$execute$1(JDBCRXGenericQueryExecutor.java:46)
        at io.vertx.reactivex.core.Vertx$3.handle(Vertx.java:625)
        at io.vertx.reactivex.core.Vertx$3.handle(Vertx.java:623)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:272)
        at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.h2.jdbc.JdbcSQLException: NULL not allowed for column "CREATED_AT"; SQL statement:
update "PUBLIC"."TBL_MODULE" set "PUBLIC"."TBL_MODULE"."SERVICE_NAME" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."MODIFIED_AT" = cast(? as timestamp), "PUBLIC"."TBL_MODULE"."CREATED_AT" = cast(? as timestamp), "PUBLIC"."TBL_MODULE"."DEPLOY_ID" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."DEPLOY_CONFIG_JSON" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."VERSION" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."SERVICE_ID" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."STATE" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."DEPLOY_LOCATION" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."SERVICE_TYPE" = cast(? as varchar), "PUBLIC"."TBL_MODULE"."PUBLISHED_BY" = cast(? as varchar) where (1 = 1 and "PUBLIC"."TBL_MODULE"."SERVICE_ID" = cast(? as varchar)) [23502-197]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:357)
        at org.h2.message.DbException.get(DbException.java:179)
        at org.h2.message.DbException.get(DbException.java:155)
        at org.h2.table.Column.validateConvertUpdateSequence(Column.java:374)
        at org.h2.table.Table.validateConvertUpdateSequence(Table.java:798)
        at org.h2.command.dml.Update.update(Update.java:157)
        at org.h2.command.CommandContainer.update(CommandContainer.java:102)
        at org.h2.command.Command.executeUpdate(Command.java:261)
        at org.h2.jdbc.JdbcPreparedStatement.execute(JdbcPreparedStatement.java:249)
        at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44)
        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java)
        at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:209)
        at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:432)
        at org.jooq.impl.AbstractDMLQuery.execute(AbstractDMLQuery.java:613)
        at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:350)
        ... 10 common frames omitted
```

--- a/core/base/src/main/java/com/nubeiot/core/statemachine/StateMachine.java
+++ b/core/base/src/main/java/com/nubeiot/core/statemachine/StateMachine.java
@@ -123,7 +123,8 @@ public final class StateMachine {
                                         ((eventAction == EventAction.PATCH || eventAction == EventAction.UPDATE) &&
                                          targetState == state))) {
             throw new StateException(
-                String.format("%s is in state %s, cannot execute action %s", objName, state, eventAction));
+                String.format("%s is in state %s, cannot execute action %s to state %s", objName, state, eventAction,
+                              targetState));
         }
     }
 

--- a/edge/core/src/main/java/com/nubeiot/edge/core/EdgeEntityHandler.java
+++ b/edge/core/src/main/java/com/nubeiot/edge/core/EdgeEntityHandler.java
@@ -31,6 +31,7 @@ import com.nubeiot.core.event.EventModel;
 import com.nubeiot.core.event.ReplyEventHandler;
 import com.nubeiot.core.exceptions.ErrorMessage;
 import com.nubeiot.core.exceptions.NotFoundException;
+import com.nubeiot.core.exceptions.NubeException;
 import com.nubeiot.core.sql.EntityHandler;
 import com.nubeiot.core.statemachine.StateMachine;
 import com.nubeiot.core.utils.DateTimes;
@@ -47,6 +48,7 @@ import com.nubeiot.edge.core.model.tables.pojos.TblTransaction;
 import com.nubeiot.edge.core.model.tables.records.TblModuleRecord;
 
 import lombok.Getter;
+import lombok.NonNull;
 
 @Getter
 public abstract class EdgeEntityHandler extends EntityHandler {
@@ -140,22 +142,19 @@ public abstract class EdgeEntityHandler extends EntityHandler {
             }
 
             ITblModule oldOne = o.orElseThrow(() -> new NotFoundException(""));
-            if (Objects.isNull(module.getState())) {
-                //re-use previous state
-                module.setState(oldOne.getState());
-            }
 
             boolean isUpdated = EventAction.UPDATE == event;
+            State targetState = Objects.isNull(module.getState()) ? oldOne.getState() : module.getState();
             if (isUpdated || EventAction.PATCH == event) {
                 return markModuleModify(module.setDeployLocation(config.getRepoConfig().getLocal()),
                                         new TblModule(oldOne), isUpdated).flatMap(
                     key -> createTransaction(key.getServiceId(), event, oldOne.toJson()).map(
-                        transId -> createPreDeployResult(key, transId, event, oldOne.getState(), module.getState())));
+                        transId -> createPreDeployResult(key, transId, event, oldOne.getState(), targetState)));
             }
             if (EventAction.REMOVE == event) {
                 return markModuleDelete(new TblModule(oldOne)).flatMap(
                     key -> createTransaction(key.getServiceId(), event, oldOne.toJson()).map(
-                        transId -> createPreDeployResult(key, transId, event, oldOne.getState(), module.getState())));
+                        transId -> createPreDeployResult(key, transId, event, oldOne.getState(), targetState)));
             }
             throw new UnsupportedOperationException("Unsupported event " + event);
         });
@@ -169,6 +168,9 @@ public abstract class EdgeEntityHandler extends EntityHandler {
                                   .prevState(prevState)
                                   .targetState(targetState)
                                   .serviceId(module.getServiceId())
+                                  .serviceFQN(module.getServiceType()
+                                                    .generateFQN(module.getServiceId(), module.getVersion(),
+                                                                 module.getServiceName()))
                                   .deployId(module.getDeployId())
                                   .deployCfg(module.getDeployConfig())
                                   .dataDir((String) this.getSharedDataFunc().apply(EdgeVerticle.SHARED_DATA_DIR))
@@ -243,9 +245,35 @@ public abstract class EdgeEntityHandler extends EntityHandler {
 
     private Single<ITblModule> markModuleModify(ITblModule module, ITblModule oldOne, boolean isUpdated) {
         logger.debug("Mark service {} to modify...", module.getServiceId());
-        ITblModule into = isUpdated ? module.into(oldOne) : oldOne.setState(module.getState());
+        ITblModule into = this.updateModule(oldOne, module, isUpdated);
         return moduleDao.update((TblModule) into.setState(State.PENDING).setModifiedAt(DateTimes.nowUTC()))
                         .map(ignore -> oldOne);
+    }
+
+    private ITblModule updateModule(@NonNull ITblModule oldOne, @NonNull ITblModule newOne, boolean isUpdated) {
+        if (Objects.nonNull(newOne.getVersion())) {
+            oldOne.setVersion(newOne.getVersion());
+        } else if (isUpdated) {
+            throw new NubeException("Version is required!");
+        }
+
+        if (Objects.nonNull(newOne.getDeployConfig())) {
+            oldOne.setDeployConfig(newOne.getDeployConfig());
+        } else if (isUpdated) {
+            throw new NubeException("Deploy config is required!");
+        }
+
+        if (Objects.nonNull(newOne.getPublishedBy())) {
+            oldOne.setPublishedBy(newOne.getPublishedBy());
+        }
+
+        if (Objects.nonNull(newOne.getState())) {
+            oldOne.setState(newOne.getState());
+        } else if (isUpdated) {
+            throw new NubeException("State is required!");
+        }
+
+        return oldOne;
     }
 
     private Single<ITblModule> markModuleDelete(ITblModule module) {
@@ -262,7 +290,7 @@ public abstract class EdgeEntityHandler extends EntityHandler {
             logger.info("Module in database is found, validate conflict ");
             StateMachine.instance()
                         .validateConflict(findModule.getState(), eventAction, "service " + findModule.getServiceId(),
-                                          targetState);
+                                          targetState == null ? findModule.getState() : targetState);
             return Optional.of(findModule);
         }
         return Optional.empty();

--- a/edge/core/src/main/java/com/nubeiot/edge/core/PreDeploymentResult.java
+++ b/edge/core/src/main/java/com/nubeiot/edge/core/PreDeploymentResult.java
@@ -43,6 +43,7 @@ public class PreDeploymentResult implements JsonData, IRequestData {
     private final State prevState;
     private final State targetState;
     private final String serviceId;
+    private final String serviceFQN;
     private final String deployId;
     private final NubeConfig deployCfg;
     @Setter
@@ -76,8 +77,8 @@ public class PreDeploymentResult implements JsonData, IRequestData {
             NubeConfig deployCfg = parseDeployCfg(this.deployCfg);
             deployCfg.setDataDir(FileUtils.recomputeDataDir(dataDir, serviceId));
             return new PreDeploymentResult(transactionId, action, Objects.isNull(prevState) ? State.NONE : prevState,
-                                           Objects.isNull(targetState) ? State.NONE : targetState,
-                                           serviceId, deployId, deployCfg, silent);
+                                           Objects.isNull(targetState) ? State.NONE : targetState, serviceId,
+                                           serviceFQN, deployId, deployCfg, silent);
         }
 
         private NubeConfig parseDeployCfg(JsonObject deployCfg) {

--- a/edge/core/src/main/java/com/nubeiot/edge/core/loader/ModuleLoader.java
+++ b/edge/core/src/main/java/com/nubeiot/edge/core/loader/ModuleLoader.java
@@ -31,10 +31,10 @@ public final class ModuleLoader implements EventHandler {
     @EventContractor(action = {EventAction.CREATE, EventAction.INIT}, returnType = Single.class)
     public Single<JsonObject> installModule(RequestData data) {
         PreDeploymentResult preResult = JsonData.from(data.getBody(), PreDeploymentResult.class);
-        logger.info("Vertx install module {}...", preResult.getServiceId());
+        logger.info("Vertx install module {}...", preResult.getServiceFQN());
         DeploymentOptions options = new DeploymentOptions(preResult.getDeployCfg().getDeployConfig()).setConfig(
             preResult.getDeployCfg().toJson());
-        return vertx.rxDeployVerticle(preResult.getServiceId(), options).doOnError(throwable -> {
+        return vertx.rxDeployVerticle(preResult.getServiceFQN(), options).doOnError(throwable -> {
             throw new EngineException(throwable);
         }).map(id -> new JsonObject().put("deploy_id", id));
     }

--- a/edge/core/src/main/java/com/nubeiot/edge/core/loader/ModuleType.java
+++ b/edge/core/src/main/java/com/nubeiot/edge/core/loader/ModuleType.java
@@ -39,27 +39,17 @@ public interface ModuleType {
 
     String name();
 
-    JsonObject serialize(@NonNull JsonObject input, @NonNull ModuleTypeRule rule) throws InvalidModuleType;
-
-    @EqualsAndHashCode
-    abstract class AbstractModuleType implements ModuleType {
-
-        @EqualsAndHashCode.Include
-        public abstract String name();
-
-        @Override
-        public final String toString() {
-            return this.name();
-        }
-
-    }
-
-
     ModuleType JAVA = new AbstractModuleType() {
 
         @Override
         public String name() {
             return "JAVA";
+        }
+
+        @Override
+        public String generateFQN(String serviceId, String version, String serviceName) {
+            return String.format("maven:%s:%s::%s", serviceId, Strings.isBlank(version) ? DEFAULT_VERSION : version,
+                                 serviceName);
         }
 
         private static final String DEFAULT_GROUP_ID = "com.nubeiot.edge.connector";
@@ -76,15 +66,13 @@ public interface ModuleType {
             if (Objects.nonNull(rule) && !rule.getRule(this).test(groupId + "." + artifactId)) {
                 throw new InvalidModuleType("Artifact is not valid");
             }
-            String serviceId = String.format("maven:%s:%s:%s::%s", groupId, artifactId,
-                                             input.getString("version", DEFAULT_VERSION), serviceName);
+            String serviceId = String.format("%s:%s", groupId, artifactId);
             return input.mergeIn(new JsonObject(), true)
                         .put("service_id", serviceId)
                         .put("service_name", serviceName)
                         .put("service_type", name());
         }
     };
-
     ModuleType JAVASCRIPT = new AbstractModuleType() {
         @Override
         public String name() {
@@ -92,11 +80,15 @@ public interface ModuleType {
         }
 
         @Override
+        public String generateFQN(String serviceId, String version, String serviceName) {
+            return null;
+        }
+
+        @Override
         public JsonObject serialize(JsonObject input, ModuleTypeRule rule) throws InvalidModuleType {
             return null;
         }
     };
-
     ModuleType GROOVY = new AbstractModuleType() {
         @Override
         public String name() {
@@ -104,11 +96,29 @@ public interface ModuleType {
         }
 
         @Override
+        public String generateFQN(String serviceId, String version, String serviceName) {
+            return null;
+        }
+
+        @Override
         public JsonObject serialize(JsonObject input, ModuleTypeRule rule) throws InvalidModuleType {
             return null;
         }
     };
 
+
+    @EqualsAndHashCode
+    abstract class AbstractModuleType implements ModuleType {
+
+        @EqualsAndHashCode.Include
+        public abstract String name();
+
+        @Override
+        public final String toString() {
+            return this.name();
+        }
+
+    }
     ModuleType SCALA = new AbstractModuleType() {
         @Override
         public String name() {
@@ -116,11 +126,15 @@ public interface ModuleType {
         }
 
         @Override
+        public String generateFQN(String serviceId, String version, String serviceName) {
+            return null;
+        }
+
+        @Override
         public JsonObject serialize(JsonObject input, ModuleTypeRule rule) throws InvalidModuleType {
             return null;
         }
     };
-
     ModuleType KOTLIN = new AbstractModuleType() {
         @Override
         public String name() {
@@ -128,11 +142,15 @@ public interface ModuleType {
         }
 
         @Override
+        public String generateFQN(String serviceId, String version, String serviceName) {
+            return null;
+        }
+
+        @Override
         public JsonObject serialize(JsonObject input, ModuleTypeRule rule) throws InvalidModuleType {
             return null;
         }
     };
-
     ModuleType RUBY = new AbstractModuleType() {
         @Override
         public String name() {
@@ -140,9 +158,19 @@ public interface ModuleType {
         }
 
         @Override
+        public String generateFQN(String serviceId, String version, String serviceName) {
+            return null;
+        }
+
+        @Override
         public JsonObject serialize(JsonObject input, ModuleTypeRule rule) throws InvalidModuleType {
             return null;
         }
     };
+
+    String generateFQN(String serviceId, String version, String serviceName);
+
+    JsonObject serialize(@NonNull JsonObject input, @NonNull ModuleTypeRule rule) throws InvalidModuleType;
+
 
 }

--- a/edge/core/src/test/java/com/nubeiot/edge/core/PreDeploymentResultTest.java
+++ b/edge/core/src/test/java/com/nubeiot/edge/core/PreDeploymentResultTest.java
@@ -22,8 +22,10 @@ public class PreDeploymentResultTest {
         PreDeploymentResult preResult = PreDeploymentResult.builder()
                                                            .transactionId("1")
                                                            .action(EventAction.REMOVE)
-                                                           .prevState(State.ENABLED).targetState(State.ENABLED)
+                                                           .prevState(State.ENABLED)
+                                                           .targetState(State.ENABLED)
                                                            .serviceId("serviceId")
+                                                           .serviceFQN("serviceFQN")
                                                            .deployId("deployId")
                                                            .deployCfg(JsonObject.mapFrom(
                                                                Collections.singletonMap("testAbc", "ab")))
@@ -33,8 +35,8 @@ public class PreDeploymentResultTest {
         preResultJson.remove("deploy_cfg");
         JSONAssert.assertEquals(
             "{\"transaction_id\":\"1\",\"action\":\"REMOVE\",\"prev_state\":\"ENABLED\",\"target_state\":\"ENABLED\"," +
-            "\"service_id\":\"serviceId\",\"deploy_id\":\"deployId\",\"silent\":false}", preResultJson.encode(),
-            JSONCompareMode.STRICT);
+            "\"service_id\":\"serviceId\",\"service_fqn\":\"serviceFQN\",\"deploy_id\":\"deployId\"," +
+            "\"silent\":false}", preResultJson.encode(), JSONCompareMode.STRICT);
         Assert.assertNotNull(deployCfg);
         Assert.assertNull(deployCfg.getSystemConfig());
         Assert.assertNotNull(deployCfg.getDataDir());
@@ -49,8 +51,10 @@ public class PreDeploymentResultTest {
         PreDeploymentResult preResult = PreDeploymentResult.builder()
                                                            .transactionId("1")
                                                            .action(EventAction.REMOVE)
-                                                           .prevState(State.ENABLED).targetState(State.ENABLED)
+                                                           .prevState(State.ENABLED)
+                                                           .targetState(State.ENABLED)
                                                            .serviceId("serviceId")
+                                                           .serviceFQN("serviceFQN")
                                                            .build();
         JsonObject preResultJson = preResult.toJson();
         preResultJson.remove("deploy_cfg");
@@ -58,7 +62,8 @@ public class PreDeploymentResultTest {
         System.out.println(deployCfg.toJson());
         JSONAssert.assertEquals(
             "{\"transaction_id\":\"1\",\"action\":\"REMOVE\",\"prev_state\":\"ENABLED\",\"target_state\":\"ENABLED\"," +
-            "\"service_id\":\"serviceId\",\"silent\":false}", preResultJson.encode(), JSONCompareMode.STRICT);
+            "\"service_id\":\"serviceId\",\"service_fqn\":\"serviceFQN\",\"silent\":false}", preResultJson.encode(),
+            JSONCompareMode.STRICT);
         Assert.assertNotNull(deployCfg);
         Assert.assertNotNull(deployCfg.getAppConfig());
         Assert.assertTrue(deployCfg.getAppConfig().isEmpty());
@@ -71,8 +76,10 @@ public class PreDeploymentResultTest {
         JsonObject jsonObject = PreDeploymentResult.builder()
                                                    .transactionId("1")
                                                    .action(EventAction.REMOVE)
-                                                   .prevState(State.ENABLED).targetState(State.ENABLED)
+                                                   .prevState(State.ENABLED)
+                                                   .targetState(State.ENABLED)
                                                    .serviceId("serviceId")
+                                                   .serviceFQN("serviceFQN")
                                                    .deployId("deployId")
                                                    .deployCfg(
                                                        JsonObject.mapFrom(Collections.singletonMap("testAbc", "ab")))
@@ -83,6 +90,7 @@ public class PreDeploymentResultTest {
         Assert.assertNotNull(preResult);
         Assert.assertEquals("1", preResult.getTransactionId());
         Assert.assertEquals("serviceId", preResult.getServiceId());
+        Assert.assertEquals("serviceFQN", preResult.getServiceFQN());
         Assert.assertEquals("deployId", preResult.getDeployId());
         Assert.assertEquals(EventAction.REMOVE, preResult.getAction());
         Assert.assertEquals(State.ENABLED, preResult.getPrevState());

